utils/frame/benchmarking-cli/src/storage/write.rs

```diff
@@ -0,0 +1,129 @@
+
+use super::{cmd::StorageCmd, record::BenchRecord};
+
+impl StorageCmd {
+	/// Benchmarks the time it takes to write a single Storage item.
+	/// Uses the latest state that is available for the given client.
+	pub(crate) fn bench_write<Block, H, C>(
+		&self,
+		cfg: &Configuration,
+		client: Arc<C>,
+		db: Arc<dyn sp_database::Database<DbHash>>,
+		storage: Arc<dyn sp_state_machine::Storage<HashFor<Block>>>,
+	) -> Result<BenchRecord>
+	where
+		Block: BlockT<Header = H, Hash = H256> + Debug,
+		H: HeaderT<Hash = H256>,
+		C: UsageProvider<Block> + HeaderBackend<Block>,
+	{
+		// Store the time that it took to write each value.
+		let mut record = BenchRecord::default();
+
+		let is_parity = matches!(cfg.database, DatabaseSource::ParityDb { path: _ });
+		let block = BlockId::Number(client.usage_info().chain.best_number);
+		let header = client.header(block)?.ok_or("Header not found")?;
+		let original_root = *header.state_root();
+		let trie = DbState::<Block>::new(storage.clone(), original_root);
+
+		info!("Preparing keys from block {}", block);
+		// Load all KV pairs and randomly shuffle them.
+		let mut kvs = trie.pairs();
+		let mut rng = Self::setup_rng();
+		kvs.shuffle(&mut rng);
+
+		info!("Writing {} keys", kvs.len());
+		// Write each value in one commit.
+		for (k, original_v) in kvs.iter() {
+			// Create a random value to overwrite with.
+			// NOTE: We use a possibly higher entropy than the original value,
+			// could be improved but acts as an over-estimation which is fine for now.
+			let mut new_v = vec![0; original_v.len()];
+			rng.fill_bytes(&mut new_v[..]);
+
+			// Interesting part here:
+			let start = Instant::now();
+			// Create a TX that will modify the Trie in the DB and
+			// calculate the root hash of the Trie after the modification.
+			let replace = vec![(k.as_ref(), Some(new_v.as_ref()))];
+			let (root, tx) = trie.storage_root(replace.iter().cloned(), self.state_version());
+			let tx = convert_tx::<Block>(tx, is_parity);
+			db.commit(tx).map_err(|e| format!("Writing to the Database: {}", e))?;
+
+			record.append(new_v.len(), start.elapsed())?;
+
+			// Now undo the change:
+			// Create a Trie with the modified root hash and replace the value with its original.
+			let trie = DbState::<Block>::new(storage.clone(), root);
+			let replace = vec![(k.as_ref(), Some(original_v.as_ref()))];
+			let (root, tx) = trie.storage_root(replace.iter().cloned(), self.state_version());
+			let tx = convert_tx::<Block>(tx, is_parity);
+			db.commit(tx).map_err(|e| format!("Writing to the Database: {}", e))?;
+
+			// Inserting the orginal value should bring us back to the original root hash.
+			if root != original_root {
+				// A crash here means hat the chain snapshot is now toast.
+				log::error!("DB corrupted. Wrong final root: {:?} vs {:?}", root, original_root);
+				std::process::exit(1);
+			}
+		}
+		Ok(record)
+	}
+}
+
+/// Converts a Trie transaction into a DB transaction.
```

Comment:
I would like an even bigger comment here explaining wtf is going on and why it is needed and what is different between parity db and rocks db.

Maybe this is not the spot to explain everything, but this is def some "hidden information" that should be documented somewhere

Reply:
I explained my understanding of it. Maybe @cheme can double-check if it is correct.

Reply:
You can probably add that this is reproducing part of the client-db crate and can easily break. Documentation should be in client-db I think.

Reply:
Actually things are different since we apply directly node removal, it is a point worth mentioning.

Reply:
Extended the comment